File: hr-backend/internal/pkg/jwt/jwt.go
// Package jwt 封装 JWT v5 的签发与解析。
//
// claims 承载 account_id、username 与过期时间（exp），用 HS256 对称签名。
// 密钥与 TTL 由 config.JWT 提供。
package jwt

import (
	"errors"
	"fmt"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

// issuer 是 token 的签发方标识，解析时强制校验，防止单服务密钥被复用时跨服务重放。
const issuer = "sili-smart-hr"

// Claims 是写入 token 的声明。
type Claims struct {
	AccountID int64  `json:"account_id"`
	Username  string `json:"username"`
	jwt.RegisteredClaims
}

// Manager 负责签发与解析 token。
type Manager struct {
	secret []byte
	ttl    time.Duration
}

// NewManager 构造一个 Manager。
func NewManager(secret string, ttl time.Duration) *Manager {
	return &Manager{secret: []byte(secret), ttl: ttl}
}

// Generate 为指定账号签发 token。
func (m *Manager) Generate(accountID int64, username string) (string, error) {
	if len(m.secret) == 0 {
		return "", errors.New("jwt: secret is empty")
	}
	now := time.Now()
	claims := Claims{
		AccountID: accountID,
		Username:  username,
		RegisteredClaims: jwt.RegisteredClaims{
			Issuer:    issuer,
			IssuedAt:  jwt.NewNumericDate(now),
			ExpiresAt: jwt.NewNumericDate(now.Add(m.ttl)),
			Subject:   username,
		},
	}
	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	return token.SignedString(m.secret)
}

// Parse 解析并校验 token，返回声明。
func (m *Manager) Parse(tokenString string) (*Claims, error) {
	var claims Claims
	_, err := jwt.ParseWithClaims(tokenString, &claims, func(t *jwt.Token) (any, error) {
		if _, ok := t.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, fmt.Errorf("jwt: unexpected signing method: %v", t.Header["alg"])
		}
		return m.secret, nil
	}, jwt.WithIssuer(issuer))
	if err != nil {
		return nil, err
	}
	return &claims, nil
}

File: hr-frontend/src/lib/jwt.ts
// 纯前端 JWT 工具：仅做本地过期判断，不做签名校验（校验在后端）。
// 用于路由守卫前置拦截过期 token，避免携带过期 token 进入受保护页面后
// 再被 401 踢回，造成登录态闪烁。

/** 解析 JWT 的 exp 声明判断是否过期。
 *  无法解析（非标准 JWT、缺 exp）时返回 false，交由后端 401 兜底。 */
export function isTokenExpired(token: string): boolean {
  const parts = token.split('.');
  if (parts.length !== 3) return false;
  try {
    const payload = JSON.parse(atob(parts[1].replace(/-/g, '+').replace(/_/g, '/'))) as { exp?: number };
    if (typeof payload.exp !== 'number') return false;
    return payload.exp * 1000 < Date.now();
  } catch {
    return false;
  }
}
